pkg/reconciler/v1alpha1/taskrun/resources/input_resource_test.go

```diff
@@ -275,13 +315,155 @@ func TestAddResourceToBuild(t *testing.T) {
 			},
 		},
 		taskRun: taskRun,
-		build:   build,
+		build:   build(),
 		wantErr: true,
 		want:    nil,
-	},
-	} {
-		setUp()
+	}, {
+		desc: "cluster resource with plain text",
```

Comment:
if you put the logic in `case v1alpha1.PipelineResourceTypeCluster:` into a separate function, you could test that function directly instead of testing it a this level

if we did that, id be happy to remove these cases completely (or just keep one at most)

Reply:
But these cases are testing two different things, 
 * `cluster resource with plain text` is validating the step is added to the build
 * `cluster resource with secrets` is validating the env from the secrets are added to the build step

Reply:
oh yeah for sure, we'd still have the 2 different tests, but we could have them at a different level.

caveat: i made a bit of a mistake and i thought this was the reconciler test! so i think what im saying is a lot less important now

however i still think this logic could be factored a bit differently so that it's slightly more clear and testable (esp as we keep adding more types of resources)

for example if you had functions like these:

```go
func getClusterStep(r *v1alpha1.PipelineResource) (*corev1.Container, error) {
	clusterResource, err := v1alpha1.NewClusterResource(r)
	if err != nil {
		return nil, fmt.Errorf("error creating cluster resource %q, resource must be invalid: %s", r.Name, err)
	}

	var envVars []corev1.EnvVar
	for _, sec := range clusterResource.Secrets {
		ev := corev1.EnvVar{
			Name: strings.ToUpper(sec.FieldName),
			ValueFrom: &corev1.EnvVarSource{
				SecretKeyRef: &corev1.SecretKeySelector{
					LocalObjectReference: corev1.LocalObjectReference{
						Name: sec.SecretName,
					},
					Key: sec.SecretKey,
				},
			},
		}
		envVars = append(envVars, ev)
	}

	clusterContainer := &corev1.Container{
		Name:  "kubeconfig",
		Image: *kubeconfigWriterImage,
		Args: []string{
			"-clusterConfig", clusterResource.String(),
		},
		Env: envVars,
	}
	return clusterContainer, nil
}

// or you could even make this functional, have it return a new build object!
func addBuildStep(b *buildv1alpha1.Build, c *corev1.Container) {
	buildSteps := append([]corev1.Container{*c}, b.Spec.Steps...)
	b.Spec.Steps = buildSteps
}
```

You could write unit tests to cover both of the above, which due to their more specific interface don't even need task runs or tasks to test.

Then inside of `AddInputResource` you could have:

```go
		case v1alpha1.PipelineResourceTypeCluster:
			{
				clusterContainlusterContainerer, err := getClusterStep(resource)
				if err != nil {
					return nil, fmt.Errorf("couldnt add cluster resource to task %q: %s", task.Name, err)
				}
				addBuildStep(build, clusterContainlusterContainerer)
			}
		}
```

And suddenly since `AddInputResource` is only calling out to 2 functions, both of which we have unit tested well, it's a lot less important (i would say not even worth it) to have test coverage at this level, and we can save ourselves a lot of boiler plate in our test code.

Reply:
I will refactor the func out of the main block of code, but would still like to keep the table test to be testing the external interface (contract) of the `AddInputResource` functionality. Testing the private function ties down the implementation. wdyt?

Reply:
I agree with @pivotal-nader-ziada thought of testing the functionality instead of implementation. Unit tests should provide the confidence to refactor code without everything completely breaking apart. Testing private functions tend to restrict the refactor scope without altering the tests.

Reply:
I see what you're saying - I think this means we should push the interface changes up a level, i.e. updated the interface to `AddInputResource` so that it doesn't need to have TaskRuns and Tasks instantiated in order to test it :D

We can keep iterating on it.

> Unit tests should provide the confidence to refactor code without everything completely breaking apart.

I dunno, I think we should limit the amount of code that one can change without also changing a unit test (and we should limit that to the ["imperative shell"](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)).

I think very finely grained and focused unit tests put pressure on us to design our code better - and I think the fact that so many objects need to be instantiated for a lot of our existing unit tests is a sign we can improve the design.

> Testing private functions tend to restrict the refactor scope without altering the tests.

I find that when we start wanting to test private functions, its often a sign that we can break up the code more, such that these private functions become public. For example, perhaps we want to have a package that is about clusters specifically, which could have a public function like `getClusterStep`

I find [this tweet from dave cheney](https://twitter.com/davecheney/status/1055046982337626112) related and inspiring:

 > use packages like a knife to separate code that consumes functionality from code that provide functionality. Use unit tests along the cut line to lock in the behaviour of your packages.